File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.cc
#include <bits/stdc++.h>

#define endl "\n"

using namespace std;

int n;

int arr[21][21] = {0, };

int Min = 999999999;

int team1[10];
int team2[10];

vector<int> t1;
vector<int> t2;


int team1sum = 0;
int team2sum = 0;

int visited[21] = { 0, };

void dfs(int cnt, int pos)
{
	if (cnt == n/2)
	{
		for (int i = 0; i < n; i++)
		{
			for (int j = 0; j < n; j++)
			{
				if (i != j && visited[i] == 1 && visited[j] == 1)
				{
					team1sum += arr[i][j];
				}
				if (i != j && visited[i] == 0 && visited[j] == 0)
				{
					team2sum += arr[i][j];
				}
			}
		}

		if (Min > abs(team1sum - team2sum))
		{
			Min = abs(team1sum - team2sum);
		}
		return;
	}

	for (int i = pos; i < n; i++)
	{
		if (!visited[i])
		{
			visited[i] = 1;
			dfs(cnt + 1, i+1);
			visited[i] = 0;
			team1sum = 0;
			team2sum = 0;
		}
	}
}


int main()
{
	cin >> n;

	for (int i = 0; i < n; i++)
	{
		for (int j = 0; j < n; j++)
		{
			cin >> arr[i][j];
		}
	}

	dfs(0, 0);

	cout << Min;

	return 0;
}

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.java
import java.util.*;
import java.io.*;

public class Main {

    static StringTokenizer st;
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringBuilder sb = new StringBuilder();

    static int n, min = Integer.MAX_VALUE;
    static int[][] map;
    static boolean[] team;

    public static void main(String[] args) throws Exception {
        n = Integer.parseInt(br.readLine());

        map = new int[n][n];
        team = new boolean[n];

        for (int i = 0 ; i < n ; i ++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0 ; j < n ; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        dfs(0, 0);

        System.out.println(min);
    }

    static void dfs(int depth, int idx) {
        if (depth == n/2) {
            int team1 = 0;
            int team2 = 0;

            for (int i = 0 ; i < n ; i ++) {
                for (int j = 0 ; j < n ; j ++) {
                    if (team[i] && team[j]) {
                        team1 += map[i][j];
                    } else if (!team[i] && !team[j]){
                        team2 += map[i][j];
                    }
                }
            }

            if (min > Math.abs(team1 - team2)) {
                min = Math.abs(team1 - team2);
            }

            return;
        }

        for (int i = idx ; i < n ; i ++) {
            if (!team[i]) {
                team[i] = true;
                dfs(depth + 1, i);
                team[i] = false;
            }
        }
    }
}